add csv and orderup text parser back in

dumbbillyhardy opened the pull request.

dumbbillyhardy committed:
add csv and orderup text parser back in

jonsmithers commented on elements/order-input.html:
```diff
@@ -67,9 +67,22 @@
-                    this._changeUrl(order);
+                    const parsersToTry = [parsers.OrderUpHtmlParser, parsers.OrderUpParser, parsers.CsvParser];
+                    const order = parsersToTry.reduce((acc, parser) => {
+                        if (acc && acc.people && acc.people.size > 0 && acc.total) {
```
There's a `order.hasPeople` getter I've been using to check for empty orders.

dumbbillyhardy replied:
That isn't completely reliable, we'd still need the total check. If I type in a csv, it parses immediately and throws an error if no number has been entered yet (so "billy,"). The total check prevents the error. Also, I think throwing errors is rather annoying, especially in the constructor but also in the "split" function. We might want to consider another signifier of a bad order parsing

jonsmithers replied:
I believe it's unreliable because csv parser uses all the deprecated `order` APIs (`withPerson`, `withTax`, `withTip`). If you fix that, you remove a lot of illegal states, and THEN, I think that checking `hasPeople` is all you need. 

I don't think it's possible to throw errors in `split()` using the correct APIs.

jonsmithers replied:
BTW, I think that the constructor and `split()` should ultimately be the same method. There's no point in having to do them separately when you always have to do them one after the other. The only reason this isn't the case is because of the deprecated `withTip`, `withTax` etc methods.

jonsmithers replied:
tl;dr: please get rid of deprecated `withPerson` `withTip` `withTax` invocations?

jonsmithers commented on elements/order-input.html:
```diff
@@ -67,9 +67,22 @@
+                        }
+                    });
+                    if (order && order.people && order.people.size > 0 && order.total) {
+                        debugger;
```
Same here. Also, debugger.

jonsmithers commented on elements/order-input.html:
```diff
@@ -67,9 +67,22 @@
-                    var order = new parsers.OrderUpHtmlParser().parse(this.$.input);
-                    console.log('order', order);
-                    this._changeUrl(order);
+                    const parsersToTry = [parsers.OrderUpHtmlParser, parsers.OrderUpParser, parsers.CsvParser];
```
Would `parsersToTry = Object.values(parsers)` work?

dumbbillyhardy replied:
That's what I originally had, but that would include the QueryStringParser, which only takes in a string, not an "elementOrText". (not a fan of the parser mutating the input, btw)

jonsmithers replied:
I don't think QueryStringParser even uses its string argument. (I also don't see where input gets mutated).

Maybe we could add a flag and do this:

```javascript
var parsersToTry = Object.values(parsers).filter(p => !p.excludeForUserInput)
```

jonsmithers commented on common/parsers.js:
```diff
@@ -144,7 +145,8 @@
     class CsvParser {
-        parse(csv) {
+        parse(elementOrText) {
+            const csv = elementOrText.innerText || elementOrText;
```
Do we ever pass in pure text? Should this just be `element`?

jonsmithers committed:
Remove deprecated method invocations

jonsmithers commented:
I gotta say....the CSV parser is really quite handy. We should figure out a way to educate users. Maybe a gif somewhere?

jonsmithers committed:
add getUserInputParsers()

jonsmithers committed:
Handle special "0" case

jonsmithers committed:
Iron out parser iteration so it works smoothly

jonsmithers committed:
Fix NaN passing validation

jonsmithers committed:
CsvParser defaults NaN to 0

jonsmithers committed:
Better yet, just ignore NaN values in CsvParser

jonsmithers committed:
Forgot to break in switch cases

jonsmithers committed:
Another special '0' case

jonsmithers committed:
Fix query string serialization bug for big numbers

jonsmithers committed:
Try html parser before csv parser

Csv parser will take just about anything

jonsmithers committed:
Handle '$' characters in csv parser

Just bringing back old functionality

jonsmithers committed:
parseFloat, not parseInt

Oops.

jonsmithers committed:
Ignore names with price of 0

This prevents premature names like 'ta', and 'fe', and 'ti', from
showing up while your typing 'tax', 'fee', or 'tip'. Not completely sure
how I feel about this. It kind of feels like the parser is harder to
play with now.

jonsmithers committed:
Chop up long line of code
